rust/bindgen.bzl

```diff
@@ -0,0 +1,156 @@
+# Copyright 2015 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def _rust_bindgen(ctx):
+    rust_toolchain = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"]
+
+    bindgen = ctx.executable.bindgen
+    rustfmt = rust_toolchain.rustfmt if hasattr(rust_toolchain, 'rustfmt') else None
+    clang = ctx.executable.clang
+    libclang = ctx.attr.libclang
+    libstdcxx = ctx.attr.libstdcxx
+    cc_toolchain = ctx.attr.cc_toolchain
+    cc_fragment = ctx.fragments.cpp
+
+    # nb. We can't grab the cc_library`s direct headers, so a header must be provided.
+    cc_lib = ctx.attr.cc_lib
+    if not hasattr(cc_lib, "cc"):
+        fail("{} is not a cc_library".format(cc_lib))
+    header = ctx.file.header
+    if header not in cc_lib.cc.transitive_headers:
+        fail("Header {} is not in {}'s transitive closure of headers.".format(ctx.attr.header, cc_lib))
+
+    # rustfmt is not in the usual place, so bindgen would fail to find it
+    bindgen_args = ["--no-rustfmt-bindings"]
+    clang_args = []
+
+    output = ctx.outputs.out
+
+    clang_libs = depset(libclang.cc.libs + libstdcxx.cc.libs)
```

Comment:
What kind of providers are these? I think the attrs need more documentation...

Reply:
.cc implies cc_* providers, will add and check (though again I am not sold on this being the best way to traverse a clang installation)

Reply:
Added the `providers = ["cc"]` requirement.